### PR TITLE
refactor(base): Remove `impl From` for `SyncTimelineEvent`

### PR DESCRIPTION
**Pull request by richvdh**
I feel like the ability to convert straight from a `Raw<AnySyncTimelineEvent>>` into a `SyncTimelineEvent` is somewhat over-simplified: the two are only occasionally equivalent, and it's better to be explicit.

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/event_item/mod.rs [hunk 1]
```diff
@@ -1023,7 +1023,7 @@ mod tests {
         formatted_body: &str,
         ts: u64,
     ) -> SyncTimelineEvent {
-        sync_timeline_event!({
+        SyncTimelineEvent::new(sync_timeline_event!({
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/event_item/mod.rs:1026]: Drive-by:constructing the `SyncTimelineEvent` should be part of the `sync_timeline_event!` macro

Comment (richvdh) [reply crates/matrix-sdk-ui/src/timeline/event_item/mod.rs:1026]: should it? I think there are cases where we actually want a `Raw<AnySyncTimelineEvent>`

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/event_item/mod.rs:1026]: It should. It's named `sync_timeline_event!` after all :)

(Long term, better, bnj-preferred fix, would be to [stop using `sync_timeline_event` and use the event factory more](https://github.com/matrix-org/matrix-rust-sdk/issues/3716), so as to consolidate testing utilities.)

Comment (richvdh) [reply crates/matrix-sdk-ui/src/timeline/event_item/mod.rs:1026]: Ok so a couple of problems with this.

First, `sync_timeline_event` is in `matrix-sdk-test`, and is used by `EventBuilder`, which is also in that crate, whereas `SyncTimelineEvent` is in `matrix-sdk-common`. We're therefore in danger of circular dependencies.

I guess we could unroll the invocations of `sync_timeline_event` which do want a `Raw<AnySyncTimelineEvent>` (such as those in `EventBuilder`) to sidestep this, but it feels inelegant.

It would also mean that `sync_timeline_event` would have quite different behaviour to `timeline_event`, `sync_state_event`, and other similar macros. Possibly we could come up with a way to change all of them, but it feels like a lot of work, and as #3716 notes, we probably want to phase out these macros than refactor half the codebase to improve them.

Maybe we instead want a *different* macro that constructs a `SyncTimelineEvent` ? I'm not sure it's worth it just to avoid the `SyncTimelineEvent::new` invocation, though.

Comment (bnjbvr) [reply crates/matrix-sdk-ui/src/timeline/event_item/mod.rs:1026]: oof, my bad, I thought that `sync_timeline_event!` did create a `SyncTimelineEvent` before this patch, and that this behavior had changed; please nevermind me.

**Review by bnjbvr — APPROVED** [review]
Thanks!
